experimental/dds/tree2/src/feature-libraries/optional-field/optionalFieldChangeFormat.ts

```diff
@@ -4,99 +4,44 @@
-import { ChangesetLocalIdSchema, EncodedChangeAtomId } from "../modular-schema";
+import { EncodedJsonableTree } from "../../core";
+import { EncodedChangeAtomId } from "../modular-schema";
 
 const noAdditionalProps: ObjectOptions = { additionalProperties: false };
 
-export const EncodedNodeUpdate = <Schema extends TSchema>(tNodeChange: Schema) =>
-	Type.Union([
-		Type.Object(
-			{
-				set: EncodedJsonableTree,
-				buildId: EncodedChangeAtomId,
-				changes: Type.Optional(tNodeChange),
-			},
-			noAdditionalProps,
-		),
-		Type.Object(
-			{
-				revert: EncodedChangeAtomId,
-				changes: Type.Optional(tNodeChange),
-			},
-			noAdditionalProps,
-		),
-	]);
+// 0 signifies "self". Using undefined doesn't actually JSON round-trip conveniently, since
+// undefined is converted to null when inside an array (which happens in e.g. the moves array).
+export const EncodedRegisterId = Type.Union([EncodedChangeAtomId, Type.Literal(0)]);
+export type EncodedRegisterId = Static<typeof EncodedRegisterId>;
 
-/**
- * Note: TS doesn't easily support extracting a generic function's return type until 4.7:
- * https://github.com/microsoft/TypeScript/pull/47607
- * This type is a workaround and can be removed once we're on a version of typescript which
- * supports expressions more like:
- * `Static<ReturnType<typeof EncodedNodeUpdate<Schema>>>`
- */
-class Wrapper<T extends TSchema> {
-	public encodedNodeUpdate(e: T) {
-		return EncodedNodeUpdate<T>(e);
-	}
-	public encodedOptionalFieldChange(e: T) {
-		return EncodedOptionalFieldChange<T>(e);
-	}
-	public encodedOptionalChangeset(e: T) {
-		return EncodedOptionalChangeset<T>(e);
-	}
-}
-
-export type EncodedNodeUpdate<Schema extends TSchema> = Static<
-	ReturnType<Wrapper<Schema>["encodedNodeUpdate"]>
->;
-
-export const EncodedOptionalFieldChange = <Schema extends TSchema>(tNodeChange: Schema) =>
-	Type.Object(
-		{
-			/**
-			 * Uniquely identifies, in the scope of the changeset, the change made to the field.
-			 * Globally unique across all changesets when paired with the changeset's revision tag.
-			 */
-			id: ChangesetLocalIdSchema,
-			/**
-			 * When populated, indicates the revision that this field change is associated with.
-			 * Is left undefined when the revision is the same as that of the whole changeset
-			 * (which would also be undefined in the case of an anonymous changeset).
-			 */
-			revision: Type.Optional(RevisionTagSchema),
-			/**
-			 * The new content for the trait. If undefined, the trait will be cleared.
-			 */
-			newContent: Type.Optional(EncodedNodeUpdate(tNodeChange)),
-			/**
-			 * Whether the field was empty in the state this change is based on.
-			 */
-			wasEmpty: Type.Boolean(),
-		},
-		noAdditionalProps,
-	);
-
-type EncodedOptionalFieldChange<Schema extends TSchema> = Static<
-	ReturnType<Wrapper<Schema>["encodedOptionalFieldChange"]>
->;
+export const EncodedBuild = Type.Object(
+	{
+		id: EncodedChangeAtomId,
+		set: EncodedJsonableTree,
```

Comment:
Could we make this a tuple for some additional savings? Doesn't have to happen in this PR though.

Reply:
might as well. There's still some more low-hanging fruit in this realm. I think encoding `self` as `null` and ChangeAtomIds without revision as just their local id is another quick win. I'll do that in a followup though.

Reply:
https://dev.azure.com/fluidframework/internal/_workitems/edit/6315